lucene/sandbox/src/java/org/apache/lucene/sandbox/codecs/quantization/KMeans.java

```diff
@@ -0,0 +1,348 @@
+      }
+      if (squaredDist < minSquaredDist) {
+        minSquaredDist = squaredDist;
+        bestCentroids = centroids;
+      }
+    }
+    return bestCentroids;
+  }
+
+  /**
+   * Initialize centroids using Forgy method: randomly select numCentroids vectors for initial
+   * centroids
+   */
+  private float[][] initializeForgy() throws IOException {
+    Set<Integer> selection = new HashSet<>();
+    while (selection.size() < numCentroids) {
+      selection.add(random.nextInt(numVectors));
+    }
+    float[][] initialCentroids = new float[numCentroids][];
+    int i = 0;
+    for (Integer selectedIdx : selection) {
+      float[] vector = vectors.vectorValue(selectedIdx);
+      initialCentroids[i++] = ArrayUtil.copyOfSubArray(vector, 0, vector.length);
+    }
+    return initialCentroids;
+  }
+
+  /** Initialize centroids using a reservoir sampling method */
+  private float[][] initializeReservoirSampling() throws IOException {
+    float[][] initialCentroids = new float[numCentroids][];
+    for (int index = 0; index < numVectors; index++) {
+      float[] vector = vectors.vectorValue(index);
+      if (index < numCentroids) {
+        initialCentroids[index] = ArrayUtil.copyOfSubArray(vector, 0, vector.length);
+      } else if (random.nextDouble() < numCentroids * (1.0 / index)) {
+        int c = random.nextInt(numCentroids);
+        initialCentroids[c] = ArrayUtil.copyOfSubArray(vector, 0, vector.length);
+      }
+    }
+    return initialCentroids;
+  }
+
+  /** Initialize centroids using Kmeans++ method */
+  private float[][] initializePlusPlus() throws IOException {
+    float[][] initialCentroids = new float[numCentroids][];
+    // Choose the first centroid uniformly at random
+    int firstIndex = random.nextInt(numVectors);
+    float[] value = vectors.vectorValue(firstIndex);
+    initialCentroids[0] = ArrayUtil.copyOfSubArray(value, 0, value.length);
+
+    // Store distances of each point to the nearest centroid
+    float[] minDistances = new float[numVectors];
+    Arrays.fill(minDistances, Float.MAX_VALUE);
+
+    // Step 2 and 3: Select remaining centroids
+    for (int i = 1; i < numCentroids; i++) {
+      // Update distances with the new centroid
+      double totalSum = 0;
+      for (int j = 0; j < numVectors; j++) {
+        // TODO: replace with RandomVectorScorer::score possible on quantized vectors
+        float dist = VectorUtil.squareDistance(vectors.vectorValue(j), initialCentroids[i - 1]);
+        if (dist < minDistances[j]) {
+          minDistances[j] = dist;
+        }
+        totalSum += minDistances[j];
+      }
+
+      // Randomly select next centroid
+      double r = totalSum * random.nextDouble();
+      double cummulativeSum = 0;
+      int nextCentroidIndex = -1;
+      for (int j = 0; j < numVectors; j++) {
+        cummulativeSum += minDistances[j];
+        if (cummulativeSum >= r) {
```

Comment:
I think there is an edge case here. `random.nextDouble()` could be 0 and the first vector could have been selected, but `0 >= 0` so this would return the same vector twice. I think you should probably do 
```suggestion
        if (cumulativeSum >= r && minDistances[j] > 0) {
```

Reply:
addressed